File: cos/olympus/src/main/java/cos/olympus/game/strategy/RespawnPlayerStrategy.java
package cos.olympus.game.strategy;

import cos.map.CreatureType;
import cos.olympus.NoSpaceException;
import cos.olympus.Util;
import cos.olympus.game.Npc;
import cos.olympus.game.Player;
import cos.olympus.game.World;
import cos.olympus.util.OpConsumer;
import cos.ops.out.Appear;


public class RespawnPlayerStrategy implements Strategy {
    private final World world;
    private final Player player;
    private final int respawnTime;

    public RespawnPlayerStrategy(int tick, World world, Player player) {
        this.world = world;
        this.player = player;
        this.respawnTime = tick + Util.rand(20, 40);
    }

    @Override
    public boolean onTick(int tick, OpConsumer outOps) {
        if (tick < respawnTime) return false;
        try {
            var cr = world.place(new Npc(player.id(), CreatureType.PLAYER, player.name()), 34, -24, 100, 1);
            outOps.add(new Appear(0, tick, cr.id(), cr.x(), cr.y(), cr.mv(), cr.sight(), cr.metrics().lvl, cr.life()));
        } catch (NoSpaceException e) {
            return false;
        }
        return true;
    }
}

File: cos/src/main/java/cos/olympus/game/strategy/RespawnPlayerStrategy.kt
package cos.olympus.game.strategy

import cos.map.ActorType.PLAYER
import cos.olympus.NoSpaceException
import cos.olympus.Util
import cos.olympus.game.Npc
import cos.olympus.game.Player
import cos.olympus.game.TickId
import cos.olympus.game.World
import cos.olympus.util.OpConsumer
import cos.ops.out.Appear

class RespawnPlayerStrategy(
    tick: TickId,
    private val world: World,
    private val player: Player
) : Strategy {

    private val respawnTime = tick + Util.rand(20, 40)

    override fun onTick(tick: TickId, outOps: OpConsumer): Boolean {
        if (tick < respawnTime) return false

        try {
            val a = world.place(Npc(player.id, PLAYER, player.name), 34, -24, 100, 1)
            outOps.add(Appear(0,  a.id, a.x, a.y, a.mv, a.sight, a.metrics.lvl, a.life))
        } catch (e: NoSpaceException) {
            return false
        }
        return true
    }
}
